client/blocks/payment-request/use-initialization.js

```diff
@@ -0,0 +1,253 @@
+	normalizeShippingOptions,
+} from '../stripe-utils';
+import { useEventHandlers } from './use-event-handlers';
+
+/**
+ * @typedef {import('../stripe-utils/type-defs').StripePaymentRequest} StripePaymentRequest
+ */
+
+export const useInitialization = ( {
+	billing,
+	shippingData,
+	setExpressPaymentError,
+	onClick,
+	onClose,
+	onSubmit,
+} ) => {
+	const stripe = useStripe();
+	/**
+	 * @type {[ StripePaymentRequest|null, function( StripePaymentRequest ):void]}
+	 */
+	// @ts-ignore
+	const [ paymentRequest, setPaymentRequest ] = useState( null );
+	const [ isFinished, setIsFinished ] = useState( false );
+	const [ isProcessing, setIsProcessing ] = useState( false );
+	const [ canMakePayment, setCanMakePayment ] = useState( false );
+	const [ paymentRequestType, setPaymentRequestType ] = useState( '' );
+	const currentShipping = useRef( shippingData );
+	const {
+		paymentRequestEventHandlers,
+		clearPaymentRequestEventHandler,
+		setPaymentRequestEventHandler,
+	} = useEventHandlers();
+
+	// Update refs when any change.
+	useEffect( () => {
+		currentShipping.current = shippingData;
+	}, [ shippingData ] );
+
+	// Create the initial paymentRequest object. Note, we can't do anything if stripe isn't available yet or we have zero total.
+	useEffect( () => {
+		if (
+			! stripe ||
+			! billing.cartTotal.value ||
+			isFinished ||
+			isProcessing ||
+			paymentRequest
+		) {
+			return;
+		}
+		const pr = getPaymentRequest( {
+			total: billing.cartTotal,
+			currencyCode: billing.currency.code.toLowerCase(),
+			countryCode: getSetting( 'baseLocation', {} )?.country,
+			shippingRequired: shippingData.needsShipping,
+			cartTotalItems: billing.cartTotalItems,
+			stripe,
+		} );
+		canDoPaymentRequest( pr ).then( ( result ) => {
+			setPaymentRequest( pr );
+			setPaymentRequestType( result.requestType || '' );
+			setCanMakePayment( result.canPay );
+		} );
+	}, [
+		billing.cartTotal,
+		billing.currency.code,
+		shippingData.needsShipping,
+		billing.cartTotalItems,
+		stripe,
+		isProcessing,
+		isFinished,
+		paymentRequest,
+	] );
+
+	// When the payment button is clicked, update the request and show it.
+	const onButtonClick = useCallback( () => {
+		setIsProcessing( true );
+		setIsFinished( false );
+		setExpressPaymentError( '' );
+		updatePaymentRequest( {
+			// @ts-ignore
+			paymentRequest,
+			total: billing.cartTotal,
+			currencyCode: billing.currency.code.toLowerCase(),
+			cartTotalItems: billing.cartTotalItems,
+		} );
+		onClick();
+	}, [
+		onClick,
+		paymentRequest,
+		setExpressPaymentError,
+		billing.cartTotal,
+		billing.currency.code,
+		billing.cartTotalItems,
+	] );
+
+	const abortPayment = useCallback( ( paymentMethod ) => {
+		paymentMethod.complete( 'fail' );
+		setIsProcessing( false );
+		setIsFinished( true );
+	}, [] );
+
+	const completePayment = useCallback( ( paymentMethod ) => {
+		paymentMethod.complete( 'success' );
+		setIsFinished( true );
+		setIsProcessing( false );
+	}, [] );
```

Comment:
I noticed a bug in testing where when 3DS fails alternate payment methods are not available because the `onClose` trigger passed in from the checkout interface isn't called when the payment method modal is closed. This is a bit fragile (but needed) logic on the part of the interface for now but until it's improved, express payment methods need to trigger this whenever their processing is "complete". The bug never surfaced until now because we never had 3DS support in the Express Payment method.

Adding the below should fix.

```suggestion
	const abortPayment = useCallback( ( paymentMethod ) => {
		paymentMethod.complete( 'fail' );
		setIsProcessing( false );
		setIsFinished( true );
		onClose();
	}, [ onClose ] );

	const completePayment = useCallback( ( paymentMethod ) => {
		paymentMethod.complete( 'success' );
		setIsFinished( true );
		setIsProcessing( false );
		onClose();
	}, [ onClose ] );
```

Reply:
Thank you! I applied the fix in f63fec4080293c14c3fa779af10972bb184ebbfb 🙂